### PR TITLE
refactor(metrics): standardize Prometheus metrics

### PR DESCRIPTION
**Pull request by lichuan0620**
**What this PR does / why we need it**:

This PR should be considered as a __breaking change__.

Refactored the metrics so that they meet the [K8 Instrumentation Guideline](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md), the changes include:

1. Removed redundant labels.
2. Added a `component` label to all metrics; it takes the value of metric namespace.
3. Renamed according to the naming convention.
4. The Histogram type request duration metric use standard bucket sizes (which produce much more data, which will put more pressure on Prometheus, but is also more precise).
5. Changed the unit of the request duration metric from milliseconds to seconds.

As a result, the metrics should now look a lot like the APIServer request metrics. 

Also, removed the `Summary` type request latency metric: `Histogram` metrics function similarly to `Summary` metrics and are generally considered as the superior option, since it supports all quantile values instead of just 0.5, 0.9, and 0.99. This should balance out the extra data volume added by the refactored Histogram metric.

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @supereagle @ddysher @iawia002 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by lichuan0620** [comment]
I'm submitting this PR just so that we can start discussing it. It does not mean that we should merge it immediately. It is a breaking change and it causes Nirvana to put more pressure on Prometheus under most circumstances (to make the Histogram metrics more precise and meaningful).

File: plugins/metrics/metrics.go [hunk 1]
```diff
@@ -45,36 +44,29 @@ type config struct {
...
 		prometheus.CounterOpts{
-			Namespace: namespace,
-			Name:      "request_count",
-			Help:      "Counter of server requests broken out for each verb, API resource, client, and HTTP response contentType and code.",
+			Namespace:   namespace,
+			Name:        "request_total",
+			Help:        "Counter of server requests broken out for each verb, API resource, and HTTP response code.",
+			ConstLabels: constLabel,
 		},
-		[]string{"method", "path", "client", "contentType", "code"},
```
Comment (ddysher) [inline plugins/metrics/metrics.go:54]: so `client` and `contentType` is not needed any more?

Comment (lichuan0620) [reply plugins/metrics/metrics.go:54]: I don't think they are important enough to justify the additional dimension. Also, I've never seen any open-sourced project adding those to their request related metrics.

**Comment by lichuan0620** [comment]
> What are the benefits of meeting Kubernetes guideline for Nirvana, apart from consistency?

Consistency by itself is a major benefit. In any production setup, one would face hundreds if not thousands of metrics. There is no way you would read the description on each and every one of them. By following the convention, you can understand the metrics by looking at their names.

For example, the first word of a metric name is usually its targeting component, and if a metric end with a `_total` postfix, then it's a counter.

If also allow you to do something like  `{__name__=~"(.+)request_duration_(.+)"}` to pull out request latency metrics from all components.

Non-naming conventions, on the other hand, are usually good for performance or/and precision. For example, the removed redundant labels save memory and make queries faster. The added Histogram buckets make `histogram_quantile` function much more predictable and precise.

> Are we actively using those Summary-based metrics now?

We are not using any Nirvana metrics right now. Even if we are, Summary is totally replaceable by Histogram so it doesn't matter. As for Kubernetes, almost all Summary-typed metrics have been depreciated as of `v1.17.0`. Other open-sourced projects are slower and more unwilling to adapt.

**Comment by iawia002** [comment]
The code LGTM, but since I'm not familiar with Prometheus, I can't give more comments about this. And if there is no more comment, I think we can merge it into the next release (0.3.x).

File: plugins/metrics/metrics.go (hunk 1 above)
Comment (ddysher) [reply plugins/metrics/metrics.go:54]: SGTM

**Comment by ddysher** [comment]
LGTM

@lichuan0620 thanks for the detailed explanation!

**Comment by supereagle** [comment]
LGTM

**Comment by ddysher** [comment]
@iawia002 make sure to make it explicit in the release note.

> This PR should be considered as a breaking change.

**Comment by iawia002** [comment]
> @iawia002 make sure to make it explicit in the release note.
> 
> > This PR should be considered as a breaking change.

Will do

**Pull request merged**